SwiftCompilerSources/Sources/Optimizer/FunctionPasses/DiagnoseInfiniteRecursion.swift

```diff
@@ -0,0 +1,578 @@
+/// infinite recursion. For example, in `f()` we can only detect the infinite recursion if we expect
+/// that the parameter `x` is invariant.
+/// ```
+///   func f(_ x: Int) {
+///     if x > 0 {   // an invariant condition!
+///       f(x)       // the call is forwarding the argument
+///     }
+///   }
+/// ```
+/// But in `g()` we can only detect the infinite recursion if we _don't_ expect that the parameter
+/// is invariant.
+/// ```
+///   func g(_ x: Int) {
+///     if x > 0 {   // no invariant condition
+///       g(x - 1)   // argument is not forwarded
+///     } else {
+///       g(x - 2)   // argument is not forwarded
+///     }
+///   }
+/// ```
+private struct Invariants: Equatable {
+
+  // Support up to 32 arguments, which should be enough in real world code.
+  // As the definition of invariants does not need to be accurate for correctness, it's fine to only support
+  // the common cases.
+  typealias ArgumentBits = UInt32
+
+  /// A bit mask of indices of arguments which are expected to be invariant.
+  /// An argument is invariant if a recursive call forwards the incoming argument.
+  /// For example:
+  /// ```
+  ///     func f(_ x: Int, _ y: Int) {
+  ///       f(x, y - 1) // The first argument is invariant, the second is not
+  ///     }
+  /// ```
+  let arguments: ArgumentBits
+
+  /// True, if all type arguments are invariant.
+  /// In contrast to `arguments` we don't distinguish between individual type arguments but have a single
+  /// flag for all type arguments.
+  /// For example:
+  /// ```
+  ///     func f<T: P>(_ t: T.Type) {
+  ///       f(T.self)   // The type argument is invariant
+  ///       f(T.V.self) // The type argument is not invariant
+  ///     }
+  /// ```
+  let typeArguments: Bool
+
+  /// True if memory content is invariant.
+  /// Like `typeArguments`, it's all or nothing. Either all memory is expected to be invariant (= never
+  /// written) or not. We could use AliasAnalysis to do a more fine-grained analysis, but in mandatory
+  /// optimizations we want to keep things simple.
+  let memory: Bool
+
+  // Nothing is invariant.
+  init() {
+    self.memory = false
+    self.typeArguments = false
+    self.arguments = 0
+  }
+
+  init(fromForwardingArgumentsOf recursiveApply: FullApplySite) {
+    let function = recursiveApply.parentFunction
+
+    // Check which parameters are exactly passed 1:1 to the recursive call.
+    var argMask: ArgumentBits = 0
+    for (argIndex, arg) in recursiveApply.arguments.enumerated() {
+      if argIndex >= MemoryLayout<ArgumentBits>.size * 8 {
+        break
+      }
+      if arg.rootValue == function.arguments[argIndex] {
+        argMask |= 1 << argIndex
+      }
+    }
+    self.arguments = argMask
+
+    // Check if the generic type parameters are exactly passed 1:1 to the recursive call.
+    let genericArgs = recursiveApply.substitutionMap.replacementTypes
```

Comment:
if you bridge `getForwardingSubstitutionMap()`, you can then compare this substitution map for equality.

Reply:
nice!